perf(lexer): tighten search loops

overlookmotel opened the pull request:
Use slice iterators in the tight search loops in `byte_search!` macro.

This shaves off 2 instructions from each turn of the loop, and makes "no match" the default path on each iteration, which is better as search spends most of its time running through bytes that *don't* match, until it find one that does.

https://godbolt.org/z/99zo8esf8

+2% on one of lexer benchmarks.

It turns out idiomatic Rust is faster than C-style pointer manipulation! Who'd have thought it?

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11118?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11118** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11118?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11118?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11116** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11116?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11118).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **May 18, 11:59 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(lexer): tighten search loops (#11118)

Use slice iterators in the tight search loops in `byte_search!` macro.

This shaves off 2 instructions from each turn of the loop, and makes "no match" the default path on each iteration, which is better as search spends most of its time running through bytes that *don't* match, until it find one that does.

https://godbolt.org/z/99zo8esf8

+2% on one of lexer benchmarks.

It turns out idiomatic Rust is faster than C-style pointer manipulation! Who'd have thought it?
